File: swiftHelper/Sources/swiftHelper/SocketMessage.swift
//
//  SocketMessage.swift
//  swiftHelper
//
//  Created by Elliot Nash on 11/16/21.
//

import Foundation

let START_OF_TEXT: UInt8 = 0x02;
let END_OF_TEXT: UInt8 = 0x03;

class SocketMessage {
    let event: String
    let data: Data
    let uuid: String
    convenience init(event: String, data: Data) {
        self.init(event: event, uuid: UUID().uuidString, data: data)
    }
    init(event: String, uuid: String, data: Data) {
        self.event = event
        self.data = data
        self.uuid = uuid
    }
    static func fromBytes(bytes: Data, bytesRead: Int) -> SocketMessage? {
        guard let eventStart = bytes.firstIndex(of: START_OF_TEXT) else {return nil}
        guard let eventEnd = bytes.firstIndex(of: END_OF_TEXT) else {return nil}
        if (bytesRead <= eventEnd+1 || eventEnd < eventStart) {return nil}
        let eventData = bytes.subdata(in: eventStart+1 ..< eventEnd)
        guard let event = String(data: eventData, encoding: .ascii) else {return nil}
                
        let remain = bytes.subdata(in: eventEnd+1 ..< bytesRead)
                
        guard let uuidStart = remain.firstIndex(of: START_OF_TEXT) else {return nil}
        guard let uuidEnd = remain.firstIndex(of: END_OF_TEXT) else {return nil}
        if (bytesRead <= uuidEnd+1 || uuidEnd < uuidStart) {return nil}
        let uuidData = remain.subdata(in: uuidStart+1 ..< uuidEnd)
        guard let uuid = String(data: uuidData, encoding: .ascii) else {return nil}
                
        let data = remain.subdata(in: uuidEnd+1 ..< remain.count)
        return SocketMessage(event: event, uuid: uuid, data: data)
    }
    func toBytes() -> Data {
        var data = Data([START_OF_TEXT])
        data.append(event.data(using: .ascii)!)
        data.append(Data([END_OF_TEXT, START_OF_TEXT]))
        data.append(uuid.data(using: .ascii)!)
        data.append(Data([END_OF_TEXT]))
        data.append(self.data)
        return data
    }
    func handleMessage() -> SocketMessage? {
        var data: Data? = nil
        if (event == "deserializeAttributedBody") {
            data = deserializeAttributedBody(data: self.data)
        }
        guard let data = data else {return nil}
        return SocketMessage(event: event, uuid: self.uuid, data: data)
    }
}

File: src/main/server/services/swiftHelperProcess/socketMessage.ts
import { generateUuid } from "@server/helpers/utils";

const START_OF_TEXT = 0x02;
const END_OF_TEXT = 0x03;

export default class SocketMessage {
    event: string;

    data: Buffer;

    uuid: string;

    constructor(event: string, data: Buffer, uuid?: string) {
        this.event = event;
        this.data = data;
        this.uuid = uuid ?? generateUuid();
    }

    static fromBytes(bytes: Buffer): SocketMessage {
        const eventStart = bytes.indexOf(START_OF_TEXT);
        const eventEnd = bytes.indexOf(END_OF_TEXT);
        const event = bytes.subarray(eventStart + 1, eventEnd).toString("ascii");
        const remain = bytes.subarray(eventEnd + 1);
        const uuidStart = remain.indexOf(START_OF_TEXT);
        const uuidEnd = remain.indexOf(END_OF_TEXT);
        const uuid = remain.subarray(uuidStart + 1, uuidEnd).toString("ascii");
        const data = remain.subarray(uuidEnd + 1);
        return new SocketMessage(event, data, uuid);
    }

    toBytes(): Buffer {
        const eventBuf = Buffer.from(this.event, "ascii");
        const uuidBuf = Buffer.from(this.uuid, "ascii");
        const buf = Buffer.alloc(eventBuf.length + 2 + uuidBuf.length + 2 + this.data.length);
        let loc = 0;
        buf[loc] = START_OF_TEXT;
        eventBuf.copy(buf, (loc += 1));
        buf[(loc += eventBuf.length)] = END_OF_TEXT;
        buf[(loc += 1)] = START_OF_TEXT;
        uuidBuf.copy(buf, (loc += 1));
        buf[(loc += uuidBuf.length)] = END_OF_TEXT;
        this.data.copy(buf, (loc += 1));
        return buf;
    }
}
